### PR TITLE
remove the usage from script in the prerequisites

### PR DESCRIPTION
**Pull request by MaggieQi**

### CODE REVIEW

**Comment by fanyangCS** [comment]
> Marketplace and Submit job v2 are designed to reproducible AI. It helps to reuse machine learning asset across projects or teams: job template sharing and reuse for docker images, data, code, job configuration and etc.

better add quotation to the two.

---
Refers to: docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md:3 in 50a427b. [](commit_id = 50a427b258f1f37b171e63672f26ab1f0aeb6c3f, deletion_comment = False)

**Comment by fanyangCS** [comment]
> In fact, OpenPAI's job configuration can be composed by tasks, docker images, script and data.

In V2,

---
Refers to: docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md:18 in 50a427b. [](commit_id = 50a427b258f1f37b171e63672f26ab1f0aeb6c3f, deletion_comment = False)

**Comment by fanyangCS** [comment]
> When writing a job configuration, you can create new components(tasks, data, script and docker images) or use existing ones.

User can compose a job configuration by creating new components or using existing ones.

---
Refers to: docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md:20 in 50a427b. [](commit_id = 50a427b258f1f37b171e63672f26ab1f0aeb6c3f, deletion_comment = False)

**Comment by fanyangCS** [comment]
> And you can share the script, data, docker images or whole job configuration on the marketplace, and other people can reuse it easily.

User can publish some components or the entire job configurations to the marketplace for other people to reuse.

---
Refers to: docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md:24 in 50a427b. [](commit_id = 50a427b258f1f37b171e63672f26ab1f0aeb6c3f, deletion_comment = False)

**Comment by fanyangCS** [comment]
> 3. Then choose a job template, for example: tensorflow_cifar10

remove

---
Refers to: docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md:66 in 50a427b. [](commit_id = 50a427b258f1f37b171e63672f26ab1f0aeb6c3f, deletion_comment = False)

File: docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md [hunk 1]
```diff
@@ -81,38 +81,226 @@ And you can share the script, data, docker images or whole job configuration on
 
 ### Introduction to yaml file
```
Comment (fanyangCS) [inline docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md]: >Introduction to yaml file [](start = 4, length = 25)

Introduction to the yaml file for job configuration

File: docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md [hunk 2]
```diff
@@ -81,38 +81,226 @@ And you can share the script, data, docker images or whole job configuration on
...
+```yaml
+protocol_version: String
+name: String
+type: job
+version: String
+contributor: String
+description: String
+retryCount: Integer
+virtualCluster: String
+gpuType: String
```
Comment (fanyangCS) [inline docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md]: >gpuType [](start = 0, length = 7)

can we rename it to co-processor?

File: docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md [hunk 3]
```diff
@@ -81,38 +81,226 @@ And you can share the script, data, docker images or whole job configuration on
...
+| `protocol_version`               | String, optional           | Protocol version, If omitted, it will be **_v2_**|
+| `name`                           | String, in `^[A-Za-z0-9\-._~]+$` format, required | Name of the prerequisite |
+| `type`                           | String, required           | Type of the prerequisite, must be **_data_**, **_script_**, or **_dockerimage_** |
+| `version`                        | String, optional           | Version of the prerequisite              |
+| `contributor`                    | String, optional           | Contributor of the prerequisite          |
+| `description`                    | String, optional           | Description of the prerequisite          |
+| `uri`                            | String, required           | Reference URL                            |
+
+
+> taskRole
```
Comment (fanyangCS) [inline docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md]: >taskRole [](start = 2, length = 8)

should be "task"?

File: docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md [hunk 4]
```diff
@@ -81,38 +81,226 @@ And you can share the script, data, docker images or whole job configuration on
...
+    data: String
+    script: String
+    dockerimage: String
+    resource: 
+      instances: Integer
+      resourcePerInstance: 
+        cpu: Integer
+        memoryMB: Integer
+        shmMB: Integer
+        gpu: Integer
```
Comment (fanyangCS) [inline docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md]: >gpu [](start = 8, length = 3)

acltr

File: docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md [hunk 5]
```diff
@@ -81,38 +81,226 @@ And you can share the script, data, docker images or whole job configuration on
...
+version: String
+contributor: String
+description: String
+retryCount: Integer
+virtualCluster: String
+gpuType: String
+
+parameters:
+  param1: valType1
+  param2: valType2
```
Comment (fanyangCS) [inline docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md]: can we add a comment, saying we can more parameters here?

File: docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md [hunk 6]
```diff
@@ -81,38 +81,226 @@ And you can share the script, data, docker images or whole job configuration on
...
+    dockerimage: String
+    resource: 
+      instances: Integer
+      resourcePerInstance: 
+        cpu: Integer
+        memoryMB: Integer
+        shmMB: Integer
+        gpu: Integer
+      portList:
+        - label: String
```
Comment (fanyangCS) [inline docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md]: >label [](start = 10, length = 5)

can we add more labels in a comment

File: docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md [hunk 7]
```diff
@@ -81,38 +81,226 @@ And you can share the script, data, docker images or whole job configuration on
...
+| `role`                           | String in `^[A-Za-z0-9._~]+$` format, required | Name for the task role, need to be unique with other roles |
+| `data`                           | String, optional           | Data that will be used in the task role, name reference of a prerequisite |
+| `script`                         | String, optional           | Script to be executed in the task role, name reference of a prerequisite |
+| `dockerimage`                    | String, optional           | Docker image to be used for the task role, name reference of a prerequisite |
+| `resource`                       | Object, required           | Resource required for the task role |
+| `minFailedTaskCount`             | Integer, optional          | Number of failed tasks to kill the entire job, null or no less than 1 |
+| `minSucceededTaskCount`          | Integer, optional          | Number of succeeded tasks to kill the entire job, null or no less than 1 |
+| `parameters`                     | Object, optional           | Specify name and value of all the referencable parameters that will be used in the task role. They can be referenced by **_$$paramName$$_**. |
```
Comment (fanyangCS) [inline docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md]: >_$$paramName$$_ [](start = 188, length = 15)

if I would like reference a parameter in job parameters section, how to do that?

File: docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md [hunk 8]
```diff
@@ -1,6 +1,6 @@
 # Marketplace and Submit job v2
 
-Marketplace and Submit job v2 are designed to reproducible AI. It helps to reuse machine learning asset across projects or teams: job template sharing and reuse for docker images, data, code, job configuration and etc.
+`Marketplace` and `Submit job v2` are designed to reproducible AI. It helps to reuse machine learning asset across projects or teams: job template sharing and reuse for docker images, data, code, job configuration and etc.
```
Comment (fanyangCS) [inline docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md]: >to [](start = 47, length = 2)

for

File: docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md [hunk 9]
```diff
@@ -79,40 +79,280 @@ And you can share the script, data, docker images or whole job configuration on
...
+virtualCluster: String
+gpuType: String
+
+parameters:
+  userParam1: valType1
+  userParam2: valType2
+  ...
+
+tasks:
+  - task
```
Comment (fanyangCS) [inline docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md]: >task [](start = 4, length = 4)

-task1
-task2
?

File: docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md [hunk 10]
```diff
@@ -79,40 +79,280 @@ And you can share the script, data, docker images or whole job configuration on
...
+protocol_version: String
+name: String
+type: dockerimage
+version: String
+contributor: String
+description: String
+uri: String
+```
+
+> `task`
```
Comment (fanyangCS) [inline docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md]: >task [](start = 3, length = 4)

task is a concept need more explanation. maybe you can add some introduction before the detailed format. 
e.g., a job may consists of multiple types of tasks, denoted by task role. Each task role container one or more containers, the instantiation of task role.

**Comment by fanyangCS** [comment]
> The system architecture is illustrated above.

can you explain the architecture in more details.

---
Refers to: docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md:400 in 1e1939b. [](commit_id = 1e1939ba021646dce5bc052f5caf93d4748afcf0, deletion_comment = False)

File: docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md [hunk 11]
```diff
@@ -79,40 +79,280 @@ And you can share the script, data, docker images or whole job configuration on
...
+    data: cifar10                     # task input data
+    script: tensorflow_cnnbenchmarks  # script executed in task
+    dockerimage: tf_example           # docker image used in task
+    resource:                         # resource required in task 
+      instances: 1                    # number of instances for the task
+      resourcePerInstance: { cpu: 2, memoryMB: 16384, gpu: 4 }
+    minSucceededTaskCount: 1          # number of succeeded tasks to kill the entire job
+    parameters:
+      data: cifar10
+      model: $$model$$
```
Comment (fanyangCS) [inline docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md]: >$$model$$ [](start = 13, length = 9)

there are multiple "model" in the spec. do we need to introduce a scope. e.g., $$parameters.model$$, $$tasks.parameters.model$$?

File: docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md [hunk 12]
```diff
@@ -79,40 +79,280 @@ And you can share the script, data, docker images or whole job configuration on
...
+      batchsize: $$batchsize$$
+    command:                          # executable commands in the task 
+      - export PYTHONPATH=$PAI_CURRENT_DIR/tensorflow_cnnbenchmarks/scripts/tf_cnn_benchmarks:$PYTHONPATH
+      - pip --quiet install scipy
+      - python tensorflow_cnnbenchmarks/scripts/tf_cnn_benchmarks/tf_cnn_benchmarks.py --local_parameter_device=gpu --variable_update=parameter_server --ps_hosts=$PAI_TASK_ROLE_ps_server_HOST_LIST --worker_hosts=$PAI_TASK_ROLE_worker_HOST_LIST --job_name=worker --task_index=$PAI_CURRENT_TASK_ROLE_CURRENT_TASK_INDEX --data_dir=$$data$$ --data_name=$$data$$ --train_dir=$PAI_OUTPUT_DIR/cifar10_output --model=$$model$$ --batch_size=$$batchsize$$
+  - role: ps_server
+    data: cifar10
+    script: tensorflow_cnnbenchmarks
+    dockerimage: tf_example
```
Comment (fanyangCS) [inline docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md]: >tf_example [](start = 17, length = 10)

likewise, $$tasks.dockerimage.dockerimage$$?

File: docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md [hunk 13]
```diff
@@ -79,40 +79,280 @@ And you can share the script, data, docker images or whole job configuration on
...
+> `data`
+
+```yaml
+protocol_version: String
+name: String
+type: data
+version: String
+contributor: String
+description: String
+uri: String
```
Comment (fanyangCS) [inline docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md]: can user specify "prerequisites" in data, script, and dockerimage?

File: docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md [hunk 14]
```diff
@@ -79,40 +79,326 @@ And you can share the script, data, docker images or whole job configuration on
...
+
+> task
+
+| Field Name                       | Schema                     | Description                              |
+| :------------------------------- | :------------------------- | :--------------------------------------- |
+| `role`                           | String in `^[A-Za-z0-9._~]+$` format, required | Name for the task, need to be unique with other roles |
+| `data`                           | String, optional           | Data that will be used in the task, name reference of a prerequisite |
+| `script`                         | String, optional           | Script to be executed in the task, name reference of a prerequisite |
+| `output`                         | String, optional           | Output that will be used in the task, name reference of a prerequisite |
+| `dockerimage`                    | String, optional           | Docker image to be used for the task, name reference of a prerequisite |
```
Comment (xudifsd) [inline docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md]: `dockerimage` is required not optional?

File: docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md [hunk 15]
```diff
@@ -79,40 +79,326 @@ And you can share the script, data, docker images or whole job configuration on
...
+type: job
+version: 1.0.0
+contributor: Alice                    # contributor of the job
+description: image classification, cifar10 dataset, tensorflow, distributed training
+retryCount: 0                         # job retry count
+virtualCluster: default               # the virtual cluster job runs on
+
+parameters:
+  model: resnet20
+  batchsize: 32
```
Comment (xudifsd) [inline docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md]: missing `data` parameter?

File: docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md (hunk 14 above)
Comment (MaggieQi) [reply docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md]: yes, dockerimage is required

File: docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md (hunk 15 above)
Comment (MaggieQi) [reply docs/marketplace-and-submit-job-v2/marketplace-and-submit-job-v2.md]: data and script are optional. If user create his dockerimage and put data and script directly into it, he can directly run the command without specify the data and script.

**Comment by fanyangCS** [comment]
closed.
